designs/dotnetcore-debugging.md

```diff
@@ -0,0 +1,196 @@
+
+Configuration process goes like that: 
+
+1. Set `request: attach` since we are attaching to already running .NET application;
+2. Set `pipeProgram` to `docker` (no need for SSH luckily);
+3. Use `exec -i <container name>` as `pipeArgs` to step inside the running container;
+4. Specify `debuggerPath` as **"/tmp/lambci_debug_files/vsdbg"** (this path reflects current mounting point for `--debugger-path` , refer to this [PR](https://github.com/awslabs/aws-sam-cli/pull/565/files) for more details);
+5. Set `quoteArgs` to `False` and let the Docker accept arguments correctly (without the need for `sh`);
+6. Set `processId` to 1 as Docker always assigns PID 1 to entry point executables.
+7. `sourceFileMap` is used for mapping breakpoints. `/var/task` is the place, where customer's code is being mounted to.
+
+_Examine sample launch.json configuration_
+
+```json
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": ".NET Core Docker Attach",
+            "type": "coreclr",
+            "request": "attach",
+            "processId": "1",
+            "pipeTransport": {
+                "pipeProgram": "docker",
+                "pipeArgs": [ "exec -i <container name>"],
+                "debuggerPath": "/tmp/lambci_debug_files/vsdbg",
+                "pipeCwd": "${workspaceRoot}",
+                "quoteArgs": false
+            },
+
+            "sourceFileMap": {
+                "/var/task": "${workspaceRoot}"
+            }
+        }
+    ]
+}
+```
+
+##### Container name caveat
+
+With this `launch.json` in place the only caveat left is the container name to step into. As for now, SAM will not assign consistent names for started containers, and neither it provides the way for the user to supply that name. So my proposal is to create _new flag_ for `sam local` command -  `--container-name` which will provide the user a way to specify the name for the container when it runs. This is a required configuration since user needs consistent naming to have usable attaching experience. SAM will use provided name and add it to `container.start()` configuration.
+
+### Open questions
+
+1. `--container-name` - does this option make sense to you, guys, maybe you have some thoughts of how to it better without exposing the ability to assign the name to the lambda container. As for me, I think this option is more than handy, and does not uncover additional implementation details, as the user is now already well aware, that his function is invoked inside Docker container. Maybe we want to have some environment variable too? Will we include that in `.samrc`? 
```

Comment:
the one thing to be aware about if we are including this in samrc is, how do we differentiate which container will be used when. Lets say, I'm debugging 3 functions at once, (3 different languages), doing a local invoke in another. All of these should work seamlessly.

If I were to assign a container name in .samrc, how would sam know which invoke maps to that container name?

Reply:
Yes, nice point, seems like this is a not an option for the `.samrc` then?

Reply:
Perhaps we could have a slightly more complex configuration section in .samrc which specifies names for each runtime's container instance? This would ultimately be the most flexible option, and should be pretty easy to implement. Thoughts?

Reply:
I think `--container-name` is a short term solution until we have a robust IDE integration which could query `docker ps` for the generated container name and connect the debugger to it. So I don't want to go too deep into this feature

Reply:
By the way, have you guys checked out [this](https://github.com/aws/aws-lambda-dotnet/tree/master/Tools/LambdaTestTool) solution? I really like how they utilized web server.

And for the auto-attach part, yes deep IDE integration could  solve the problem

Reply:
No container name flag is required any more. Resolving this.